### ISSUE TITLE
much higher CPU load with LS5 compared to LS2

### ISSUE THREAD

**Issue by skrueger8** [elastic/logstash#6634]
Hi,

we're using Logstash to capture syslog events and send them to a graylog server using the lumberjack plugin. We're using LS 2.4.1 right now but we want to upgrade to LS5.

During testing we found out that LS5 has a much higher CPU usage even when idling, here's a grafana load usage from an idle server (on the left is CPU load with LS2, on the right is LS5)

![ls5_load](https://cloud.githubusercontent.com/assets/18461169/22584768/72dc64e4-e9f4-11e6-969a-87b9c539cdfc.png)

- Version:

Logtash 2.4.1 is fine, Logstash 5.x is not

- Operating System:

Solaris 11.3/SPARC

```
# pkg list entire
NAME (PUBLISHER)                                  VERSION                    IFO
entire    
```                                        0.5.11-0.175.3.16.0.3.0    i--

# java -version
java version "1.8.0_121"
Java(TM) SE Runtime Environment (build 1.8.0_121-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.121-b13, mixed mode)

# pargs 2356
2356:   /usr/bin/java -XX:+UseParNewGC -XX:+UseConcMarkSweepGC -XX:CMSInitiatingOccupan
argv[0]: /usr/bin/java
argv[1]: -XX:+UseParNewGC
argv[2]: -XX:+UseConcMarkSweepGC
argv[3]: -XX:CMSInitiatingOccupancyFraction=75
argv[4]: -XX:+UseCMSInitiatingOccupancyOnly
argv[5]: -XX:+UseLargePages
argv[6]: -XX:ParallelGCThreads=4
argv[7]: -XX:+DisableExplicitGC
argv[8]: -Djava.awt.headless=true
argv[9]: -Dfile.encoding=UTF-8
argv[10]: -XX:+HeapDumpOnOutOfMemoryError
argv[11]: -Xmx1g
argv[12]: -Xms256m
argv[13]: -Xss2048k
argv[14]: -Djffi.boot.library.path=.../vendor/jruby/lib/jni
argv[15]: -Xbootclasspath/a:.../vendor/jruby/lib/jruby.jar
argv[16]: -classpath
argv[17]: :
argv[18]: -Djruby.home=.../vendor/jruby
argv[19]: -Djruby.lib=.../vendor/jruby/lib
argv[20]: -Djruby.script=jruby
argv[21]: -Djruby.shell=/bin/sh
argv[22]: org.jruby.Main
argv[23]: .../lib/bootstrap/environment.rb
argv[24]: logstash/runner.rb
argv[25]: -f
argv[26]: .../config/logstash.conf
argv[27]: -w
argv[29]: 4
```

- Config File (if you have sensitive info, please remove it):

jvm.options
```
## optimizations
-XX:+UseLargePages
-XX:ParallelGCThreads=4
```

logstash.conf
```
input {
  file {
    path           => "/var/adm/messages"
    type           => "messages"
    start_position => "beginning"
    sincedb_path   => ".../data/var_adm_messages"
  }
}

filter {
  mutate {
    add_field => [ "system" , "solaris11" ]
    add_field => [ "mandant" , "nix" ]
  }
}

output {
  lumberjack {
    codec           => "json"
    hosts           => "graylog"
    port            => 6378
    ssl_certificate => ".../config/ca-cert.pem"
  }
}
```

- Sample Data:

Hot methods seem to be:

```
# ./j_cputime.d -p 2356
...
   2356 method   org/jruby/internal/runtime/methods/JittedMethod.post    10339
   2356 method   org/jruby/internal/runtime/methods/JittedMethod.call    10351
   2356 method   org/jruby/internal/runtime/methods/JittedMethod.pre     10351
   2356 method   org/jruby/internal/runtime/methods/DynamicMethod.getImplementationClass    10361
   2356 method   org/jruby/runtime/Arity.checkArity                      10362
   2356 method   org/jruby/internal/runtime/methods/JittedMethod.getArity    10381
   2356 method   org/jruby/runtime/Arity.isFixed                         10520
   2356 method   org/jruby/runtime/Arity.required                        10520
   2356 method   org/jruby/RubyArray.safeArrayRef                        10672
   2356 method   org/jruby/RubyBasicObject.<init>                        11448
   2356 method   org/jruby/RubyObject.<init>                             11448
   2356 method   org/jruby/ast/executable/RuntimeCache.getValue          11725
   2356 method   org/jruby/runtime/ivars/VariableAccessor.get            11927
   2356 method   org/jruby/runtime/ivars/VariableAccessor.getVariable    11927
   2356 method   org/jruby/RubyModule.hashCode                           12147
   2356 method   org/jruby/runtime/ivars/VariableAccessor.getClassId     13323
   2356 method   java/lang/Thread.currentThread                          13449
   2356 method   java/nio/HeapCharBuffer.ix                              15101
   2356 method   java/nio/HeapCharBuffer.put                             15101
   2356 method   java/nio/Buffer.nextPutIndex                            15283
   2356 method   java/lang/String.charAt                                 16670
   2356 method   java/lang/ref/Reference.get                             17364
   2356 method   org/jruby/RubyClass.getRealClass                        19043
   2356 method   org/jruby/RubyBasicObject.getRuntime                    21107
   2356 method   org/jruby/RubyClass.getClassRuntime                     21223
   2356 method   org/jruby/runtime/callsite/CachingCallSite.call         22830
   2356 method   org/jruby/runtime/ThreadContext.callThreadPoll          25679
   2356 method   org/jruby/runtime/callsite/CachingCallSite.pollAndGetClass    25679
   2356 method   org/jruby/runtime/callsite/CacheEntry.typeOk            31351
   2356 method   org/jruby/RubyModule.getGeneration                      31835
   2356 method   java/lang/Object.<init>                                 42085
   2356 method   org/jruby/RubyBasicObject.getMetaClass                  69301
```

**Comment by jsvd** [comment]
Logstash 5 does indeed do more even when idle than 2.x used to do, mainly because of the metric collection subsystem.
That said, part of this load may be related to https://github.com/elastic/logstash/pull/6705, that caused frequent thread dumps unnecessarily. This PR has been merged and should be included in upcoming 5.2.2 and 5.3.0

**Comment by skrueger8** [comment]
Looks much better when the fix from #6705 is applied to LS 5.2.1

Old Logstash 2.4.1 load is 0.10
patched Logstash 5.2.1 load is 0.15 (was 0.5 before!)

Thanks.

**Comment by jsvd** [comment]
Thanks for confirming @skrueger8, closing this issue
